### ISSUE TITLE
APP CRASH (c6474fda6ad mpirun) drcachesim failed to write trace in offline mode

### ISSUE THREAD

**Issue by varun729** [DynamoRIO/dynamorio#4892]
**Describe the bug**
drcachesim failed to write trace in offline mode. When running drcachesim without "-offline", application runs fine and dumps the cache simulation results.

I have a patch (see Additional Context) that solves this problem. I don't know if it is the correct solution, so I have included it.

**To Reproduce**
1. Minimal source code: See the test_mpi.c application on https://usc-rc.github.io/tutorials/open-mpi 
2. Compiled MPICH v3.0.4 from source (./configure; make; make install)repository
3. Compile and run
<pre>
```
mpicc test_mpi.c -o test_mpi
dynamorio/exports/bin64/drrun -t drcachesim -- mpirun -np 1 ./test_mpi   # runs fine without -offline
dynamorio/exports/bin64/drrun -t drcachesim -offline -- mpirun -np 1 ./test_mpi   # failed to write trace in offline mode
```
</pre>

Please also answer these questions:
 - What happens when you run without any client?
   - runs fine 
 - What happens when you run with debug build
   - throws the error "Failed to create a subdir in ."

**Expected behavior**
Application completes and generates traces in the application directory drmemory.test_mpi.id.num.dir (id is pid or tid, and num is a number < 10000)

**Screenshots or Pasted Text**
Output (obfuscated the full path) with failure
<pre>
```
WARNING: cannot find .../dynamorio/exports/bin64/../lib64/debug/libdynamorio.so: is this an incomplete installation?
WARNING: .../dynamorio/exports/bin64/.. does not appear to be a valid DynamoRIO root
Failed to create a subdir in .

===================================================================================
=   BAD TERMINATION OF ONE OF YOUR APPLICATION PROCESSES
=   EXIT CODE: 255
=   CLEANING UP REMAINING PROCESSES
=   YOU CAN IGNORE THE BELOW CLEANUP MESSAGES
===================================================================================
```
</pre>

**Versions**
 - What version of DynamoRIO are you using?
   - latest c6474fda6ad (need support for AVX512 so cannot use a released version)
 - Does the latest build from https://github.com/DynamoRIO/dynamorio/releases solve the problem?
   - no
 - What operating system version are you running on? ("Windows 10" is *not* sufficient: give the release number.)
   - Linux 5.4.0-72-generic (Ubuntu 18.04.5 LTS)
 - Is your application 32-bit or 64-bit?
   - 64-bit

**Additional context**
Did some debugging on my own. Use the following patch to see that the application name contains the full path instead of only the basename of the executable.

Debugging patch:
<pre>
```
diff --git a/clients/drcachesim/tracer/tracer.cpp b/clients/drcachesim/tracer/tracer.cpp
index c7d02bee..39f1b4fd 100644
--- a/clients/drcachesim/tracer/tracer.cpp
+++ b/clients/drcachesim/tracer/tracer.cpp
@@ -1873,8 +1873,10 @@ init_offline_dir(void)
         if (file_ops_func.create_dir(buf))
             break;
     }
-    if (i == NUM_OF_TRIES)
+    if (i == NUM_OF_TRIES) {
+        FATAL("Failed here 1 : %s : %d : %s : %s : %s\n", op_outdir.get_value().c_str(), dr_get_process_id(), subdir_prefix, dr_get_application_name(), buf);
         return false;
+    }
     /* We group the raw thread files in a further subdir to isolate from the
      * processed trace file.
      */
```
</pre>

Patch that resolves the problem:
<pre>
diff --git a/core/unix/os.c b/core/unix/os.c
index ec89c366..db9babb5 100644
--- a/core/unix/os.c
+++ b/core/unix/os.c
@@ -1094,7 +1094,7 @@ get_application_name(void)
 DYNAMORIO_EXPORT const char *
 get_application_short_name(void)
 {
-    return get_application_name_helper(false, false /* short name */);
+    return get_application_name_helper(true, false /* short name */);
 }

 /* Sets pointers to the application's command-line arguments. These pointers are then used
</pre>

Output with patch:
<pre>
```
WARNING: cannot find .../dynamorio/exports/bin64/../lib64/debug/libdynamorio.so: is this an incomplete installation?
WARNING: .../dynamorio/exports/bin64/.. does not appear to be a valid DynamoRIO root
Hello world from processor machinename, rank 0 out of 1 processors
```
</pre>

**Comment by derekbruening** [comment]
dr_get_application_name(), as its documentation shows, is supposed to return only the basename, and changing it to have path separators in it would cause all kinds of failures I would think: that is not what we want, nor would we change get_application_short_name() which as its own name says is supposed to be the short (==basename) version without any path.

I don't understand what the underlying problem is: what is the directory that it tried to create?  I guess that's never printed out: what did your "Failed here 1" line print?  How is this application different from others where we have never seen this issue -- Is this happening in a forked child only where the CWD has changed (-outdir default is "."), or in the top-level process?

**Comment by varun729** [comment]
The patch does not change the behavior of dr_get_application_name(). The patch disables the cache, so the basename of the application is calculated in the call to function get_application_name_helper.

The debug patch results in the following output:
<pre>
```
WARNING: cannot find .../dynamorio/exports/bin64/../lib64/debug/libdynamorio.so: is this an incomplete installation?
WARNING: .../dynamorio/exports/bin64/.. does not appear to be a valid DynamoRIO root
Failed here 1 : . : 11589 : drmemtrace : irname/mpi_test/test_mpi : ./drmemtrace.irname/mpi_test/test_mpi.11589.6624.dir

===================================================================================
=   BAD TERMINATION OF ONE OF YOUR APPLICATION PROCESSES
=   EXIT CODE: 255
=   CLEANING UP REMAINING PROCESSES
=   YOU CAN IGNORE THE BELOW CLEANUP MESSAGES
===================================================================================
```
</pre>

From the "Failed here 1" line, you can see that the output of dr_get_application_name() is "irname/mpi_test/test_mpi". For your information, the full path of the executable is "/dirname/mpi_test/test_mpi".
The expected value is "test_mpi".

**Comment by derekbruening** [comment]
> The patch does not change the behavior of dr_get_application_name().
OK, this is what was confusing as it seemed like it was talking about the patch to os.c:
> Did some debugging on my own. Use the following patch to see that the application name contains the full path instead of only the basename of the executable.

But I don't understand what caused the problem, nor how passing ignore_cache=true would help in any case: with -early_inject (the default and only supported config; non-early is deprecated and will be removed at some point) there is no cache, so this parameter (and thus the os.c patch here) should be a nop, right?

I can't reproduce any problem running drcachesim offline on apps that fork and exec with paths like "./foo".

**Comment by derekbruening** [comment]
So I did find a reproducer:

```
$ cp suite/tests/bin/common.fib .
$ bin64/drrun -t drcachesim -offline -- suite/tests/bin/linux.execve64 ./common.fib 
parent is running under DynamoRIO
parent waiting for child
child is running under DynamoRIO
Failed to create a subdir in .
child has exited
```

But a better understanding of what is happening is needed.  Re-computing the basename from the full path every single call does not seem the right solution, nor does it make sense to me why that would make a difference.

**Comment by varun729** [comment]
The problem is not about the paths "./foo". It has to do with mpirun. You can notice that I am running with a single MPI rank here for reproducing the problem. If I run the application as below, then I can collect the trace without any issue.
<pre>
```
dynamorio/exports/bin64/drrun -t drcachesim -offline -- ./test_mpi
```
</pre>

**Comment by varun729** [comment]
I agree that re-computing the basename from the full path every single call is the right solution, which is why I created this issue to get inputs from DynamoRIO team.

Thanks @derekbruening for the reproducer. This decouples the problem from MPI.

**Comment by derekbruening** [comment]
This is a regression from the fix for #1660 where the later setting of the executable_path fails to re-compute the basename pointer into that buffer.

**Comment by varun729** [comment]
Interesting. Runs fine without the patch if I use the absolute path of the executable.
